core/libp2p/connection/yamux/yamux_stream.cpp

```diff
@@ -0,0 +1,303 @@
+      }
+      return yamuxed_connection_->streamAckBytes(
+          stream_id_, bytes,
+          [self{shared_from_this()}, cb = std::move(cb), bytes](auto &&res) {
+            self->is_reading_ = false;
+            if (!res) {
+              return cb(res.error());
+            }
+            self->read_buffer_.consume(bytes);
+            cb(bytes);
+          });
+    }
+
+    // else, set a callback, which is called each time a new data arrives
+    yamuxed_connection_->streamAddDataNotifyee(
+        stream_id_,
+        [self{shared_from_this()}, cb = std::move(cb), out, bytes]() mutable {
+          if (self->read_buffer_.size() < bytes) {
+            // not yet
+            return false;
+          }
+
+          if (boost::asio::buffer_copy(boost::asio::buffer(out.data(), bytes),
+                                       self->read_buffer_.data(), bytes)
+              != bytes) {
+            cb(Error::INTERNAL_ERROR);
+          } else {
+            self->yamuxed_connection_->streamAckBytes(
+                self->stream_id_, bytes,
+                [self, cb = std::move(cb), out, bytes](auto &&res) {
+                  (void)out;
+                  self->is_reading_ = false;
+                  if (!res) {
+                    return cb(res.error());
+                  }
+                  self->read_buffer_.consume(bytes);
+                  cb(bytes);
+                });
+          }
+          return true;
+        });
+  }
+
+  void YamuxStream::readSome(gsl::span<uint8_t> out, size_t bytes,
+                             ReadCallbackFunc cb) {
+    if (bytes == 0 || out.empty() || static_cast<size_t>(out.size()) < bytes) {
+      return cb(Error::INVALID_ARGUMENT);
+    }
+    if (!is_readable_) {
+      return cb(Error::NOT_READABLE);
+    }
+    if (is_reading_) {
+      return cb(Error::IS_READING);
+    }
+
+    is_reading_ = true;
+
+    // if there is already enough data in our buffer, read it immediately
+    if (read_buffer_.size() != 0) {
+      auto to_read = std::min(read_buffer_.size(), bytes);
+      if (boost::asio::buffer_copy(boost::asio::buffer(out.data(), to_read),
+                                   read_buffer_.data(), to_read)
+          != to_read) {
+        return cb(Error::INTERNAL_ERROR);
+      }
+      return yamuxed_connection_->streamAckBytes(
+          stream_id_, to_read,
+          [self{shared_from_this()}, cb = std::move(cb), to_read](auto &&res) {
+            self->is_reading_ = false;
+            if (!res) {
+              return cb(res.error());
+            }
+            self->read_buffer_.consume(to_read);
+            cb(to_read);
+          });
+    }
+
+    // else, set a callback, which is called each time a new data arrives
+    yamuxed_connection_->streamAddDataNotifyee(
+        stream_id_,
+        [self{shared_from_this()}, cb = std::move(cb), out, bytes]() mutable {
+          if (self->read_buffer_.size() == 0) {
+            // not yet
+            return false;
```

Comment:
why this callback returns bool?

Reply:
This is done to signalize to YamuxedConnection, that this notifyee should be removed from the list. The lambdas can be called several time until there's really enough bytes to be read/written